### PR TITLE
zql [5/n] Add integration / e2e tests for ZQL

### PR DESCRIPTION
**Pull request by tantaman**
don't review. Putting up as a draft so I can link to it now.

### CODE REVIEW

File: src/zql/integration.test.ts [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+  expect(
+    'This test fails because `experimentalWatch` does not call us with an empty array when we want initial data from an empty collection. So we wait for forever for data to be available.',
+  ).toEqual('');
+  const {q} = setup();
+  const rows = await q.select('id').prepare().exec();
+  expect(rows).toEqual([]);
+});
+
+test('prepare a query before the collection has writes then run it', async () => {
+  // This fails since it appears that experimentalWatch is not calling
```
Comment (tantaman) [inline src/zql/integration.test.ts:97]: incorrect comment. This is the case for the test above, not this one.

File: src/zql/integration.test.ts [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+}
+
+// why is it backwards?
+const compareIds = (a: {id: string}, b: {id: string}) =>
+  a.id < b.id ? 1 : a.id > b.id ? -1 : 0;
+
+const compareIdsAsc = (a: {id: string}, b: {id: string}) =>
+  a.id < b.id ? -1 : a.id > b.id ? 1 : 0;
+
+// This test fails because `experimentalWatch` does not call us with an empty array when we want initial data from an empty collection.
```
Comment (grgbkr) [inline src/zql/integration.test.ts:85]: Fun bug.

Comment (arv) [reply src/zql/integration.test.ts:116]: Seems to be working correctly. Something else is going on here.

File: src/zql/integration.test.ts [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+
+test('subscribing to a query calls us with the complete query results on change', async () => {
+  const issues = sampleTenUniqueIssues();
+  const {q, r} = setup();
+  await Promise.all(issues.map(r.mutate.initIssue));
+
+  let called: (v: unknown) => void;
+  const calledPromise = new Promise(resolve => {
+    called = resolve;
+  });
```
Comment (arv) [inline src/zql/integration.test.ts:176]:
```suggestion
  let resolve: (v: unknown) => void;
  const calledPromise = new Promise(res => {
    resolve = res;
  });
```

File: src/zql/integration.test.ts (hunk 2 above)
Comment (tantaman) [reply src/zql/integration.test.ts:116]: I found some tests for this case and they pass. I also added some tests here for it (spying on experimental watch) and those work too. Not sure what's going on yet.

File: src/zql/integration.test.ts [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+  title: fc.string(),
+  status: fc.constantFrom('open', 'closed'),
+  priority: fc.constantFrom('high', 'medium', 'low'),
+  assignee: fc.string(),
+  created: fc.date(),
+  updated: fc.date(),
+  closed: fc.option(fc.date(), {nil: undefined}),
+});
+
+// const operators = ['=', '<', '>', '>=', '<=', 'IN', 'LIKE', 'ILIKE'] as const;
```
[the remark this answers is not in the document]
Comment (tantaman) [reply src/zql/integration.test.ts]: Nope. I'll add those

**Comment by tantaman** [comment]
going ahead and merging since this contains bug fixes.

Two failing tests had to be commented out until I can debug what is preventing 1-shot queries from working against empty collections.

**Review by arv — COMMENTED** [review]
LGTM

**Pull request merged**